### PR TITLE
Added an argument to specify the parts directory for ReadsSparkSink

### LINKED ISSUE #4636: Add Support To ReadsSparkSink to specify parts directory
Currently there is no way to specify where the bam parts get written to before being combined into a single sorted bam. Being able to specify this directory is important for supporting google compute engine LOCAL disks which currently can't be stitched into a logical volume.

### PR DESCRIPTION
**Pull request by jamesemery**
This should allow for the .parts whiteout to live elsewhere which is needed for using LOCAL disks with spark. 

Fixes #4636

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java [hunk 1]
```diff
@@ -180,7 +180,7 @@ private void writeBam(final JavaRDD<GATKRead> reads, final String inputBamPath,
         try {
             ReadsSparkSink.writeReads(ctx, outputPath, bwaArgs.referencePath, reads, header,
                     shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE,
-                    PSUtils.pathseqGetRecommendedNumReducers(inputBamPath, numReducers, getTargetPartitionSize()));
+                    PSUtils.pathseqGetRecommendedNumReducers(inputBamPath, numReducers, getTargetPartitionSize()), null);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqBwaSpark.java]: I think these nulls should all be replaced with references to `shardedPartsDir`

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqPipelineSpark.java [hunk 2]
```diff
@@ -289,7 +289,7 @@ protected void runTool(final JavaSparkContext ctx) {
                 final int numPartitions = Math.max(1, (int) (numTotalReads / readsPerPartitionOutput));
                 final JavaRDD<GATKRead> readsFinalRepartitioned = readsFinal.coalesce(numPartitions, false);
                 ReadsSparkSink.writeReads(ctx, outputPath, null, readsFinalRepartitioned, header,
-                        shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE, numPartitions);
+                        shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE, numPartitions, null);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqPipelineSpark.java]: likewise

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqScoreSpark.java [hunk 3]
```diff
@@ -221,7 +221,7 @@ protected void runTool(final JavaSparkContext ctx) {
         if (outputPath != null) {
             try {
                 ReadsSparkSink.writeReads(ctx, outputPath, null, readsFinal, header,
-                        shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE, recommendedNumReducers);
+                        shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE, recommendedNumReducers, null);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqScoreSpark.java]: same here

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/BwaAndMarkDuplicatesPipelineSpark.java [hunk 4]
```diff
@@ -66,7 +66,7 @@ protected void runTool(final JavaSparkContext ctx) {
                         referenceArguments.getReferencePath().toAbsolutePath().toUri().toString(),
                         markedReads, bwaEngine.getHeader(),
                         shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE,
-                        getRecommendedNumReducers());
+                        getRecommendedNumReducers(), null);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/BwaAndMarkDuplicatesPipelineSpark.java]: and here

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java [hunk 5]
```diff
@@ -87,14 +89,35 @@ private void shutdownMiniCluster() {
...
+        final File outputFile = createTempFile(outputFileName, outputFileExtension);
+        final File outputDir = createTempDir(outputFileName + ".someOtherPlace");
+
+        // Make a directory with unusable permissions in place of where the default file will live
+        final File defaultDir = new File(outputFile.getAbsolutePath() + ".parts/");
+        defaultDir.mkdir();
+        BufferedWriter testfile = new BufferedWriter(new FileWriter(defaultDir.getAbsolutePath() + "/test.txt"));
+        testfile.write("test");
+        testfile.close();
+        Runtime.getRuntime().exec("chmod a-w -R " + defaultDir.getAbsolutePath()+"/");
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java]: IntelliJ makes a big fuss about how `defaultDir.setWriteable(false)` will not be respected in the test suite, and i found that it wasn't resulting in failure when we tried to write to it.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java [hunk 6]
```diff
@@ -228,10 +228,10 @@ private static void saveAsShardedHadoopFiles(
 
     private static void writeReadsSingle(
             final JavaSparkContext ctx, final String outputFile, final String referenceFile, final SAMFormat samOutputFormat, final JavaRDD<SAMRecord> reads,
-            final SAMFileHeader header, final int numReducers) throws IOException {
+            final SAMFileHeader header, final int numReducers, final String outputPartsDir) throws IOException {
 
         final JavaRDD<SAMRecord> sortedReads = SparkUtils.sortReads(reads, header, numReducers);
-        final String outputPartsDirectory = outputFile + ".parts/";
+        final String outputPartsDirectory = outputPartsDir==null?outputFile + ".parts/":outputPartsDir;
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java]: done

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java [hunk 7]
```diff
@@ -87,14 +89,35 @@ private void shutdownMiniCluster() {
...
+    }
+
+    @Test(dataProvider = "loadReadsBAM", groups = "spark")
+    public void outputDirectoryTest(String inputBam, String outputFileName, String referenceFile, String outputFileExtension) throws IOException {
+        final File outputFile = createTempFile(outputFileName, outputFileExtension);
+        final File outputDir = createTempDir(outputFileName + ".someOtherPlace");
+
+        // Make a directory with unusable permissions in place of where the default file will live
+        final File defaultDir = new File(outputFile.getAbsolutePath() + ".parts/");
+        defaultDir.mkdir();
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java [hunk 8]
```diff
@@ -87,14 +89,35 @@ private void shutdownMiniCluster() {
...
+        BufferedWriter testfile = new BufferedWriter(new FileWriter(defaultDir.getAbsolutePath() + "/test.txt"));
+        testfile.write("test");
+        testfile.close();
+        Runtime.getRuntime().exec("chmod a-w -R " + defaultDir.getAbsolutePath()+"/");
+
+        assertSingleShardedWritingWorks(inputBam, referenceFile, outputFile.getAbsolutePath(), outputDir.getAbsolutePath());
+
+        // Test that the file wasn't deleted when spark cleared its temp directory
+        Assert.assertEquals(new File(defaultDir.getAbsolutePath() + "/test.txt").exists(), true);
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java [hunk 9]
```diff
@@ -87,14 +89,35 @@ private void shutdownMiniCluster() {
...
+        assertSingleShardedWritingWorks(inputBam, referenceFile, outputFile.getAbsolutePath(), null);
+    }
+
+    @Test(dataProvider = "loadReadsBAM", groups = "spark")
+    public void outputDirectoryTest(String inputBam, String outputFileName, String referenceFile, String outputFileExtension) throws IOException {
+        final File outputFile = createTempFile(outputFileName, outputFileExtension);
+        final File outputDir = createTempDir(outputFileName + ".someOtherPlace");
+
+        // Make a directory with unusable permissions in place of where the default file will live
+        final File defaultDir = new File(outputFile.getAbsolutePath() + ".parts/");
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java]: done

File: src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java [hunk 10]
```diff
@@ -66,6 +66,7 @@
     public static final String BAM_PARTITION_SIZE_LONG_NAME = "bam-partition-size";
     public static final String NUM_REDUCERS_LONG_NAME = "num-reducers";
     public static final String SHARDED_OUTPUT_LONG_NAME = "sharded-output";
+    public static final String OUTPUT_SHARD_DIR_LONG_NAME = "output-shard-dir";
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java]: done

File: src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java [hunk 11]
```diff
@@ -91,6 +92,11 @@
             optional = true)
     protected boolean shardedOutput = false;
 
+    @Argument(doc = "For tools that write an output, directory for writing sharded chunks before being combined",
+            fullName = OUTPUT_SHARD_DIR_LONG_NAME,
+            optional = true)
+    protected String shardedPartsDir = null;
+
     @Argument(doc="For tools that shuffle data or write an output, sets the number of reducers. Defaults to 0, which gives one partition per 10MB of input.",
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java]: I added a check in the writer which will throw.

**Comment by jamesemery** [comment]
Responded to a round of comments

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java [hunk 12]
```diff
@@ -87,14 +89,35 @@ private void shutdownMiniCluster() {
     @Test(dataProvider = "loadReadsBAM", groups = "spark")
     public void readsSinkTest(String inputBam, String outputFileName, String referenceFile, String outputFileExtension) throws IOException {
         final File outputFile = createTempFile(outputFileName, outputFileExtension);
-        assertSingleShardedWritingWorks(inputBam, referenceFile, outputFile.getAbsolutePath());
+        assertSingleShardedWritingWorks(inputBam, referenceFile, outputFile.getAbsolutePath(), null);
+    }
+
+    @Test(dataProvider = "loadReadsBAM", groups = "spark")
+    public void outputDirectoryTest(String inputBam, String outputFileName, String referenceFile, String outputFileExtension) throws IOException {
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java]: I rewrote this:

```
    @Test(dataProvider = "loadReadsBAM", groups = "spark")
    public void testSpecifyPartsDir(String inputBam, String outputFileName, String referenceFile, String outputFileExtension) throws IOException {
        final File outputFile = createTempFile(outputFileName, outputFileExtension);
        final File nonDefaultShardsDir = createTempDir(outputFileName + ".someOtherPlace");


        final java.nio.file.Path defaultPartsDir = IOUtils.getPath(ReadsSparkSink.getDefaultPartsDirectory(outputFile.getAbsolutePath()));
        final java.nio.file.Path subpath = defaultPartsDir.resolve("subpath");

        // Make a directory with unusable permissions in place of where the default parts file will live
        try {
            final Set<PosixFilePermission> readOnly = EnumSet.of(PosixFilePermission.OWNER_READ);
            final FileAttribute<Set<PosixFilePermission>> readOnlyPermissions = PosixFilePermissions.asFileAttribute(readOnly);

            Files.createDirectory(defaultPartsDir);
            // An empty directory seems to be able to be deleted even with write permissions disabled, so put a file in it
            Files.createFile(subpath, readOnlyPermissions);
            Files.setPosixFilePermissions(defaultPartsDir, readOnly);

            //assert it fails when writing to the default path
            Assert.assertThrows(() -> assertSingleShardedWritingWorks(inputBam, referenceFile, outputFile.getAbsolutePath(), null));

            //show this succeeds when specifying a different path for the parts directory
            assertSingleShardedWritingWorks(inputBam, referenceFile, outputFile.getAbsolutePath(), nonDefaultShardsDir.getAbsolutePath());

            // Test that the file wasn't deleted when spark cleared its temp directory
            Assert.assertTrue(Files.exists(defaultPartsDir));
        } finally {
            try {
                final EnumSet<PosixFilePermission> readWriteExecute = EnumSet.of(PosixFilePermission.OWNER_READ,
                                                                          PosixFilePermission.OWNER_WRITE,
                                                                          PosixFilePermission.OWNER_EXECUTE);
                Files.setPosixFilePermissions(defaultPartsDir, readWriteExecute);
                Files.setPosixFilePermissions(subpath, readWriteExecute);
                Files.deleteIfExists(subpath);
                Files.deleteIfExists(defaultPartsDir);
            } catch (IOException e) {
                System.out.print("Failed to delete test file");
                e.printStackTrace();
            }
        }
    }
```

Comment (lbergelson) [reply src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSinkUnitTest.java]: @jamesemery I changed it to be more assertive and also to not shell out.

**Comment by lbergelson** [comment]
@jamesemery Apparently my test doesn't work on travis.  Gaaah.

**Comment by jamesemery** [comment]
@lbergelson Returned to shelling out.

**Comment by lbergelson** [comment]
so sad

**Pull request merged**